core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java

```diff
@@ -53,14 +56,43 @@ public static Collection<PartitionStats> computeStats(Table table, Snapshot snap
 
     StructType partitionType = Partitioning.partitionType(table);
     List<ManifestFile> manifests = snapshot.allManifests(table.io());
-    Queue<PartitionMap<PartitionStats>> statsByManifest = Queues.newConcurrentLinkedQueue();
-    Tasks.foreach(manifests)
-        .stopOnFailure()
-        .throwFailureWhenFinished()
-        .executeWith(ThreadPools.getWorkerPool())
-        .run(manifest -> statsByManifest.add(collectStats(table, manifest, partitionType)));
+    return collectStats(table, manifests, partitionType).values();
+  }
+
+  /**
+   * Computes the partition stats incrementally after the given snapshot to current snapshot.
+   *
+   * @param table the table for which partition stats to be computed.
+   * @param afterSnapshot the snapshot after which partition stats is computed (exclusive).
+   * @param currentSnapshot the snapshot till which partition stats is computed (inclusive).
+   * @return the {@link PartitionMap} of {@link PartitionStats}
+   */
+  public static PartitionMap<PartitionStats> computeStatsIncremental(
+      Table table, Snapshot afterSnapshot, Snapshot currentSnapshot) {
+    Preconditions.checkArgument(table != null, "Table cannot be null");
+    Preconditions.checkArgument(Partitioning.isPartitioned(table), "Table must be partitioned");
+    Preconditions.checkArgument(currentSnapshot != null, "Current snapshot cannot be null");
+    Preconditions.checkArgument(afterSnapshot != null, "Snapshot cannot be null");
+    Preconditions.checkArgument(currentSnapshot != afterSnapshot, "Both the snapshots are same");
+    Preconditions.checkArgument(
+        SnapshotUtil.isAncestorOf(table, currentSnapshot.snapshotId(), afterSnapshot.snapshotId()),
+        "Starting snapshot %s is not an ancestor of current snapshot %s",
+        afterSnapshot.snapshotId(),
+        currentSnapshot.snapshotId());
 
-    return mergeStats(statsByManifest, table.specs());
+    Set<Long> snapshotIdsRange =
+        Sets.newHashSet(
+            SnapshotUtil.ancestorIdsBetween(
+                currentSnapshot.snapshotId(), afterSnapshot.snapshotId(), table::snapshot));
+    StructType partitionType = Partitioning.partitionType(table);
+    List<ManifestFile> manifests =
+        currentSnapshot.allManifests(table.io()).stream()
+            .filter(
+                manifestFile ->
+                    snapshotIdsRange.contains(manifestFile.snapshotId())
+                        && !manifestFile.hasExistingFiles())
```

Comment:
Note that not including the existing records manifest for incremental compute as those stats will be previously written when they are added.